File: typescript/src/solve.ts
import { parseArgs } from 'node:util';
import { parse, parseFromFile } from './parser.js';
import { solveWithLogic } from './solver.js';
import { Board } from './primitives/board.js';
import { printWithStats } from './printer.js';

const solve = () => {
  const { values } = parseArgs({
    options: {
      'puzzle-path': { type: 'string' },
      puzzle: { type: 'string' },
    },
  });

  let board: Board;

  if (values['puzzle-path'] !== undefined && values.puzzle !== undefined) {
    console.error('Choose either --puzzle-path or --puzzle, not both!');
    process.exit(1);
  } else if (values['puzzle-path'] !== undefined) {
    board = parseFromFile(values['puzzle-path']);
  } else if (values.puzzle !== undefined) {
    board = parse(values.puzzle);
  } else {
    console.error(
      'Usage: ' +
        'npx tsx src/solve.ts --puzzle-path <path-to-puzzle.json> ' +
        'OR ' +
        'npx tsx src/solve.ts --puzzle <raw puzzle json> '
    );
    process.exit(1);
  }

  board.initializeCandidates();

  const startingCandidateCount = board.getCandidateCount();
  const startingTime = Date.now();
  const solvedBoard = solveWithLogic(board).board;
  console.log(
    printWithStats(solvedBoard, startingTime, startingCandidateCount)
  );
};

solve();

File: kotlin/src/main/kotlin/Solve.kt
package dev.patientallison.sudoku

fun main(args: Array<String>) {
    val puzzlePath = parseArg(args, "--puzzle-path")
    val puzzle = parseArg(args, "--puzzle")

    val board = when {
        puzzlePath != null && puzzle != null -> {
            System.err.println("Choose either --puzzle-path or --puzzle, not both!")
            return
        }
        puzzlePath != null -> parseFromFile(puzzlePath)
        puzzle != null -> parse(puzzle)
        else -> {
            System.err.println("Usage: --puzzle-path <path> OR --puzzle <json>")
            return
        }
    }

    board.initializeCandidates()
    val startTime = System.nanoTime()
    solveWithLogic(board).board
    val elapsed = System.nanoTime() - startTime
    println("Board is solved!")
    println("Time: ${elapsed / 1_000_000}ms")
}

private fun parseArg(args: Array<String>, key: String): String? =
    args.indexOf(key).let { if (it >= 0) args.getOrNull(it + 1) else null }

